### PR TITLE
Refine filter

### PR DESCRIPTION
**Pull request by nick-pestell**
This PR implements a new filter design where `SNP` is required to have a min_read_pth of 8 reads supporting the `ALT`, i.e. `AD1>=8` AND a min allele frequency of 0.8. For a `REF`, there must be at least 1 read, and a majority supporting the `REF`. All sites not meeting the criteria for either a SNP or REF are returned as Ns in the consensus file, are omitted from the snps_tab and the filtered_bcf.bcf.

This design is implemented in `mask.bash`, by using `bcftools filter` to find the sites that do not meet the requirements for a `REF` or `SNP` and adding these sites to the rpt mask.

`SNP`s distinguished from `REF`s in `vcf2consensus.bash` which also uses `bcftools filter`  with the same SNP criteria as above.

It also updates two unit test: 

* mask_test.py
* vcf2consensus_test.py

The general pattern for these two tests is that a hypothetical genome of 13 bases long witha  reference of all `A`s is tested. A test VCF file, `edge-cases.vcf` is used as input. The output `masked.bed` from `mask.bash` is compared to the ideal `edge-cases.bed`. The output consensus file from `vcf2consensus.bash` is compared to the ideal `edge-cases.fas`. These two tests do not rely on each other, i.e. `vcf2consensus_test.py` can pass even if `mask_test.py` fails.

### CODE REVIEW

File: bTB-WGS_process.nf [hunk 1]
```diff
@@ -152,7 +153,8 @@ process VarCall {
 	tuple pair_id, file("mapped.bam") from mapped_bam
 
 	output:
-	tuple pair_id, file("${pair_id}.vcf.gz") into vcf, vcf2
+	tuple pair_id, file("${pair_id}.vcf.gz"), file("${pair_id}.vcf.gz.csi") into vcf, vcf2,
+	vcf4mask
```
Comment (ellisrichardj) [inline bTB-WGS_process.nf:157]: Would be best to have this on a single line for readability

File: bTB-WGS_process.nf [hunk 2]
```diff
@@ -198,15 +200,16 @@ process VCF2Consensus {
 	maxForks 2
 
 	input:
-	tuple pair_id, file("mask.bed"), file("variant.vcf.gz") from vcf_bed
+	tuple pair_id, file("mask.bed"), file("nonmasked-regions.bed"), file("variant.vcf.gz"),
+	file("variant.vcf.gz.csi") from vcf_bed
```
Comment (ellisrichardj) [inline bTB-WGS_process.nf:204]: Would be best to have this on a single line for readability

File: bin/mask.bash [hunk 3]
```diff
@@ -11,23 +11,35 @@
...
+allsites=$5
+MIN_READ_DEPTH=$6
+MIN_ALLELE_FREQUENCY=$7
 
-# Find zero coverage regions
-bedtools genomecov -bga -ibam $bam |
-grep -w "0\$" | 
-cat > zero_cov.bed
+# Construct a mask: 
+# mask regions which don't have {sufficient evidence for alt AND suffecient evidence for the REF} OR {zero coverage}
```
Comment (ellisrichardj) [inline bin/mask.bash:29]: Please correct spelling of 2nd sufficient

File: bin/mask.bash [hunk 4]
```diff
@@ -11,23 +11,35 @@
...
+    (ALT!='.' && INFO/AD[1]/(INFO/AD[0]+INFO/AD[1]) < ${MIN_ALLELE_FREQUENCY} && INFO/AD[1] >= INFO/AD[0]) ||
+    (ALT='.' && AD=0)" $vcf -ov -o quality-mask.vcf
+bedtools merge -i quality-mask.vcf > quality-mask.bed
 
-# Mask repeat regions
-cat zero_cov.bed $rpt_mask | 
+# Merge with exisiting known repeat regions
+cat quality-mask.bed $rpt_mask | 
 sort -k1,1 -k2,2n |
+cut -f4 --complement |
```
Comment (ellisrichardj) [inline bin/mask.bash:38]: This is a good fix - the alternative is to remove the final column in the repeat mask file - its not needed for anything else.  It was needed for compatibility with the output from bedtools genomecov - which isn't used now

File: bin/vcf2Consensus.bash [hunk 5]
```diff
@@ -10,38 +10,39 @@
 #%
 #% INPUTS
 #%    ref          input path to reference genome (.fas)
-#%    bed          input path to mask (.bed)
+#%    mask         input path to mask (.bed)
+#%    regions      input path to regions file containing all regions
+#%                 for masking/filtering(.bed)
```
Comment (ellisrichardj) [inline bin/vcf2Consensus.bash:15]: Would be best to have this on a single line for readability

File: tests/data/edge-cases.vcf [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+##INFO=<ID=DP4,Number=4,Type=Integer,Description="Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases">
+##INFO=<ID=MQ,Number=1,Type=Integer,Description="Average mapping quality">
+##bcftools_callVersion=1.10.2+htslib-1.10.2
+##bcftools_callCommand=call -mf GQ -Ou -; Date=Thu Jul 15 11:18:52 2021
+##bcftools_normVersion=1.10.2+htslib-1.10.2
+##bcftools_normCommand=norm -f ./references/Mycbovis-2122-97_LT708304.fas -Oz -o /tmp/tmpk0th1bgd/variants.vcf.gz -; Date=Thu Jul 15 11:18:52 2021
+##bcftools_viewVersion=1.10.2+htslib-1.10.2
+##bcftools_viewCommand=view -Oz -o tinyvariants.vcf.gz tinyvariants.vcf; Date=Thu Oct 14 17:25:39 2021
+#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	/tmp/tmpk0th1bgd//tinymatch.bam
+LT708304-Mycobacteriumbovis-AF2122-97	1	.	A	.	59	.	DP=1;AD=1;MQ0F=0;AN=2;DP4=1,0,0,0;MQ=60	GT	0/0
```
Comment (ellisrichardj) [inline tests/data/edge-cases.vcf:38]: A

**Review by ellisrichardj — APPROVED** [review]
All looks good - thanks Nick.   Need to think about proportions to confirm a ref too.  A simple majority seems to simple, and I think an N might actually be more appropriate where there's similar evidence for REF and ALT
